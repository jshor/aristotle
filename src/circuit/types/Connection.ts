import CircuitNode from '../base/CircuitNode'

class Connection {
  node: CircuitNode
  id: string

  constructor (node: CircuitNode, id: string) {
    this.node = node
    this.id = id
  }
}

export default Connection
