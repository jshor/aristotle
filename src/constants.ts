/** Minimum zoom percentage. */
export const MIN_ZOOM = 0.1

/** Maximum zoom percentage. */
export const MAX_ZOOM = 2

/** The incremental percentage amount per zoom operation. */
export const ZOOM_STEP = 0.1

/** The frictional coefficient for animated panning movements. */
export const PANNING_FRICTION = 4

/** The speed scalar for animated panning movements using momentum. */
export const PANNING_SPEED = 15

/** The amount of padding, in pixels, for an exported image. */
export const IMAGE_PADDING = 20

/** The amount of momentum to build with each key press while moving an object using arrow keys. */
export const ARROW_KEY_MOMENTUM_MULTIPLIER = 1.05

/** The amount of time, in milliseconds, that the user must be touching something before emitting a `touchhold` event. */
export const TOUCH_HOLD_TIMEOUT = 100

/**
 * The minimum CSS z-index value to apply to all items.
 * This high number is to ensure that all items visually overlap all wires.
 */
export const ITEM_BASE_Z_INDEX = 1000

/** The easing function for each animation frame to ease out after panning. */
export const PANNING_EASING_FUNCTION = (x: number) => 1 - Math.pow(1 - x, 3)
