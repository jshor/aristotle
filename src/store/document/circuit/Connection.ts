import CircuitNode from './CircuitNode'

class Connection {
  node: CircuitNode
  id: string

  constructor (node: CircuitNode, id: string) {
    this.node = node
    this.id = id
  }
}

export default Connection
