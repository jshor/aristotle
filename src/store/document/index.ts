import { StoreGetters, defineStore } from 'pinia'
import { DocumentState, state } from './state'
import * as clipboardActions from './actions/clipboard'
import * as connectionActions from './actions/connection'
import * as editorActions from './actions/editor'
import * as groupingActions from './actions/grouping'
import * as itemActions from './actions/item'
import * as panningActions from './actions/panning'
import * as portActions from './actions/port'
import * as positioningActions from './actions/positioning'
import * as oscillatorActions from './actions/oscillator'
import * as rotationActions from './actions/rotation'
import * as selectionActions from './actions/selection'
import * as simulationActions from './actions/simulation'
import * as sizingActions from './actions/sizing'
import * as snappingActions from './actions/snapping'
import * as undoRedoActions from './actions/undoRedo'
import * as zIndexActions from './actions/zIndex'
import * as zoomingActions from './actions/zooming'
import * as getters from './getters'
import sortByZIndex from '@/utils/sortByZIndex'
import BaseItem from '@/types/interfaces/BaseItem'
import Item from '@/types/interfaces/Item'
import Connection from '@/types/interfaces/Connection'

export const createDocumentStore = (documentId: string) => defineStore({
  id: documentId,
  state,
  getters: {
    ...getters,

    baseItems (state) {
      return (Object
        .values(state.connections) as BaseItem[])
        .concat(Object.values(state.items) as BaseItem[])
        .sort(sortByZIndex)
    },

    hasSelection (state) {
      return state.selectedItemIds.size > 0 || state.selectedConnectionIds.size > 0
    },

    hasSelectedItems (state) {
      return state.selectedItemIds.size > 0
    },

    canGroup (state) {
      let ungroupedCount = 0
      let groupedCount = 0

      const hasUngrouped = (list: Set<string>, map: Record<string, Connection | Item>) => {
        for (const id of list) {
          map[id].groupId
            ? groupedCount++
            : ungroupedCount++

          if (ungroupedCount > 1 || (groupedCount && ungroupedCount)) {
            // at least two items are ungrouped, or at least one item is grouped and at least one item is ungrouped
            return true
          }
        }

        return false
      }

      return hasUngrouped(state.selectedItemIds, state.items) || hasUngrouped(state.selectedConnectionIds, state.connections)
    },

    canUngroup (state) {
      return state.selectedGroupIds.size > 0
    },

    canUndo (state) {
      return state.undoStack.length > 0
    },

    canRedo (state) {
      return state.redoStack.length > 0
    }
  },
  actions: {
    ...clipboardActions,
    ...connectionActions,
    ...editorActions,
    ...groupingActions,
    ...itemActions,
    ...panningActions,
    ...portActions,
    ...positioningActions,
    ...oscillatorActions,
    ...rotationActions,
    ...selectionActions,
    ...simulationActions,
    ...snappingActions,
    ...sizingActions,
    ...undoRedoActions,
    ...zIndexActions,
    ...zoomingActions
  }
})

export type DocumentStore = ReturnType<typeof createDocumentStore>

export type DocumentStoreInstance = ReturnType<DocumentStore>
