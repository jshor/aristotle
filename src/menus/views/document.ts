import { MenuFactory } from '@/types/interfaces/MenuFactory'
import { DocumentStore } from '@/store/document'
import { createHelpSubmenu } from '../submenus/help'
import { createEditSubmenu } from '../submenus/edit'
import { createFileSubmenu } from '../submenus/file'
import { createViewSubmenu } from '../submenus/view'
import { t } from '@/utils/i18n'

/**
 * Creates the document application menu.
 */
export const createDocumentMenu: MenuFactory = (useDocumentStore?: DocumentStore) => {
  return [
    {
      label: t('menu.file.parent'),
      submenu: createFileSubmenu(useDocumentStore)
    },
    {
      label: t('menu.edit.parent'),
      submenu: createEditSubmenu(useDocumentStore)
    },
    {
      label: t('menu.view.parent'),
      submenu: createViewSubmenu(useDocumentStore)
    },
    {
      label: t('menu.help.parent'),
      role: 'help',
      submenu: createHelpSubmenu()
    }
  ]
}
