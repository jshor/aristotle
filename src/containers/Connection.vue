<template>
  <draggable
    :style="{ zIndex }"
    :position="position"
    :is-selected="isSelected"
    :allow-touch-drag="freeportId !== null"
    @touchhold="onDragStart"
    @drag="onDrag"
    @drag-end="onDragEnd"
    @select="k => selectItem(id, k)"
    @deselect="deselectItem(id)"
    @contextmenu="onContextMenu"
  >
    <wire
      :source="source"
      :target="target"
      :geometry="geometry"
      :is-selected="isSelected"
      :is-preview="isPreview"
      :flash="flash"
      :label="'TODO'"
      ref="root"
      data-test="wire"
    />
  </draggable>
</template>

<script lang="ts">
import { v4 as uuid } from 'uuid'
import { ComponentPublicInstance, defineComponent, PropType, ref, computed } from 'vue'
import Draggable from '@/components/interactive/Draggable.vue'
import Wire from '@/components/Wire.vue'
import { DocumentStore } from '@/store/document'
import boundaries from '@/store/document/geometry/boundaries'
import renderLayout from '@/store/document/geometry/wire'
import editorContextMenu from '@/menus/context/editor'

export default defineComponent({
  name: 'Connection',
  components: {
    Draggable,
    Wire
  },
  props: {
    /** Whether or not this connection is selected. */
    isSelected: {
      type: Boolean,
      default: false
    },

    /** ID of the connection. */
    id: {
      type: String,
      required: true
    },

    /** CSS z-index value. */
    zIndex: {
      type: Number,
      default: 0
    },

    /** Whether or not this connection is just for previewing. */
    isPreview: {
      type: Boolean,
      default: false
    },

    /** Whether or not the item should show a flash once to the user. */
    flash: {
      type: Boolean,
      default: false
    },

    /** Document store instance. */
    store: {
      type: Function as PropType<DocumentStore>,
      required: true
    }
  },
  setup (props) {
    const store = props.store()
    const connection = ref(store.connections[props.id])
    const root = ref<ComponentPublicInstance<HTMLElement>>()
    const source = computed(() => store.ports[store.connections[props.id].source])
    const target = computed(() => store.ports[store.connections[props.id].target])
    const geometry = computed(() => renderLayout(source.value, target.value))
    const position = computed(() => {
      const topLeft = boundaries.getExtremePoint('min', source.value.position, target.value.position)

      return {
        y: topLeft.y + geometry.value.minY,
        x: topLeft.x + geometry.value.minX
      }
    })

    const freeportId = ref<string | null>(null)

    /**
     * Mouse button down event handler.
     *
     * This will inform the component that the mouse is down and ready to create a new freeport, if it moves.
     */
    function onDragStart ({ touches }: TouchEvent) {
      // TODO: bug: if the cursor drags only 1px (or some insufficient amount to create a new freeport) then the connection will inadvertantly be deleted
      createFreeport({
        x: touches[0].clientX,
        y: touches[0].clientY
      })
    }

    function createFreeport (position: Point) {
      if (!root.value?.$el || connection.value.groupId !== null) return

      // directly change the DOM style (as opposed to using v-show/refs) to avoid "flickering" while VDOM updates
      root.value.$el.style.opacity = '0'
      freeportId.value = uuid()
      navigator.vibrate(20)

      store.createFreeport({
        itemId: freeportId.value,
        inputPortId: uuid(),
        outputPortId: uuid(),
        connectionChainId: connection.value.connectionChainId,
        sourceId: source.value.id,
        targetId: target.value.id,
        value: source.value.value
      })

      store.setSnapBoundaries(freeportId.value)
      store.dragItem(freeportId.value, position)
    }

    /**
     * Mouse move event handler.
     *
     * If the mouse is held down on the wire, then movement in both the x and y directions
     * will cause a new freeport to be created in that position.
     *
     * This will only create a new freeport once per mousedown-move cycle.
     */
    async function onDrag (position: Point) {
      if (!freeportId.value) {
        createFreeport(position)
      } else {
        store.dragItem(freeportId.value, position)
      }
    }

    /**
     * Mouse up event handler.
     */
    function onDragEnd () {
      if (freeportId) {
        // after dragging has finished, two new connections are established
        // this connection is now ready to destroy itself
        store.destroyConnection({
          source: connection.value.source,
          target: connection.value.target
        })
      }

      freeportId.value = null
    }

    function onContextMenu () {
      if (!props.isSelected) {
        store.selectItem(props.id)
      }

      window.api.showContextMenu(editorContextMenu(props.store))
    }

    return {
      root,
      geometry,
      position,
      onDragStart,
      onDrag,
      onDragEnd,
      onContextMenu,
      source,
      target,
      freeportId,
      selectItem: store.selectItem,
      deselectItem: store.deselectItem
    }
  }
})
</script>
