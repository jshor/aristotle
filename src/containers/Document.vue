<template>
  <editor
    ref="editorRef"
    :zoom="store.zoom"
    :grid-size="gridSize"
    :tabindex="0"
    :canvas="store.canvas"
    @pan="store.panDelta"
    @zoom="store.setZoom"
    @contextmenu="onContextMenu"
    @resize="store.setViewerSize"
  >
    <selector
      :zoom="store.zoom"
      @selection-end="store.createSelection"
      @selection-start="onDeselect"
      ref="selector"
    />

    <template
      v-if="store.hasLoaded"
      v-for="baseItem in store.baseItems"
    >
      <item
        v-if="'type' in baseItem"
        :store="storeDefinition"
        :id="baseItem.id"
        :key="baseItem.id"
        :is-selected="baseItem.isSelected"
        :flash="flash"
        :z-index="baseItem.zIndex + ITEM_BASE_Z_INDEX"
      />
      <!-- Note: z-index of items will be offset by ITEM_BASE_Z_INDEX to ensure they overlap wires -->

      <connection
        v-else
        :store="storeDefinition"
        :id="baseItem.id"
        :key="`c${baseItem.id}`"
        :is-selected="baseItem.isSelected"
        :flash="flash"
        :z-index="baseItem.zIndex"
      />
    </template>

    <port-item
      v-for="(port, id) in store.ports"
      :key="id"
      :port="port"
      :store="storeDefinition"
    />

    <group
      v-for="group in store.groups"
      :key="group.id"
      :id="group.id"
      :bounding-box="group.boundingBox"
      :is-selected="group.isSelected"
      :z-index="store.zIndex"
      :zoom="store.zoom"
    />
  </editor>
</template>

<script lang="ts">
import { defineComponent, PropType, onBeforeUnmount, ref, watch, computed, onMounted } from 'vue'
import Connection from './Connection.vue'
import Item from './Item.vue'
import PortItem from './PortItem.vue'
import Editor from '@/components/editor/Editor.vue'
import Selector from '@/components/editor/Selector.vue'
import Group from '@/components/Group.vue'
import boundaries from '@/store/document/geometry/boundaries'
import printing from '@/utils/printing'
import { DocumentStore } from '@/store/document'
import { useRootStore } from '@/store/root'
import { usePreferencesStore } from '@/store/preferences'
import { ARROW_KEY_MOMENTUM_MULTIPLIER, IMAGE_PADDING, ITEM_BASE_Z_INDEX, PRINTER_FRIENDLY_COLORS } from '@/constants'
import BoundingBox from '@/types/types/BoundingBox'
import Point from '@/types/interfaces/Point'
import { DocumentStatus } from '@/types/enums/DocumentStatus'
import { ViewType } from '@/types/enums/ViewType'
import { createEditSubmenu } from '@/menus/submenus/edit'

export default defineComponent({
  name: 'Document',
  components: {
    Group,
    Editor,
    Selector,
    Connection,
    Item,
    PortItem
  },
  props: {
    store: {
      type: Function as PropType<DocumentStore>,
      required: true
    }
  },
  setup (props) {
    const store = props.store()
    const rootStore = useRootStore()
    const preferencesStore = usePreferencesStore()
    const editorRef = ref<typeof Editor>()
    const updates = ref(0)
    const flash = computed(() => store.isDebugging && store.isCircuitEvaluated)
    const gridSize = computed(() => {
      return preferencesStore.grid.showGrid.value
        ? preferencesStore.grid.gridSize.value
        : 0
    })

    let acceleration = 1
    let keydownAnimationFrameId = 0
    let clipboardAnimationFrameId = 0

    // add event listeners when the document is viewed, or remove them otherwise
    watch(() => rootStore.viewType, viewType => {
      if (viewType === ViewType.Document) {
        onDocumentReady()
      } else {
        offDocumentReady()
      }
    }, { immediate: true })

    watch(() => store.status, onDocumentStatus, { immediate: true })

    onBeforeUnmount(offDocumentReady)

    /**
     * Handles the requested document status.
     */
    function onDocumentStatus (status: DocumentStatus) {
      switch (status) {
        case DocumentStatus.Printing:
          return printImage()
        case DocumentStatus.SavingImage:
          return exportImage()
      }
    }

    /**
     * Adds clipboard and keyboard event listeners.
     */
    function onDocumentReady () {
      document.addEventListener('cut', onCut)
      document.addEventListener('copy', onCopy)
      document.addEventListener('paste', onPaste)
      document.addEventListener('keydown', onKeyDown)
      document.addEventListener('keyup', onKeyUp)
      window.addEventListener('blur', onKeyUp)
    }

    /**
     * Removes clipboard and keyboard event listeners.
     */
    function offDocumentReady () {
      document.removeEventListener('cut', onCut)
      document.removeEventListener('copy', onCopy)
      document.removeEventListener('paste', onPaste)
      document.removeEventListener('keydown', onKeyDown)
      document.removeEventListener('keyup', onKeyUp)
      window.removeEventListener('blur', onKeyUp)
    }

    /**
     * Cuts the selection from this document.
     */
    function onCut ($event: ClipboardEvent) {
      return onClipboard('cut', $event)
    }

    /**
     * Copies the document selection to the clipboard.
     */
    function onCopy ($event: ClipboardEvent) {
      return onClipboard('copy', $event)
    }

    /**
     * Pastes from the clipboard into the document.
     */
    function onPaste ($event: ClipboardEvent) {
      return onClipboard('paste', $event)
    }

    /**
     * Clipboard event handler.
     */
    function onClipboard (action: 'cut' | 'copy' | 'paste', $event: ClipboardEvent) {
      if (!($event.target instanceof HTMLInputElement)) {
        if (clipboardAnimationFrameId) return

        clipboardAnimationFrameId = requestAnimationFrame(() => {
          clipboardAnimationFrameId = 0
          store[action]()
          cancelAnimationFrame(clipboardAnimationFrameId)
        })
      }

      rootStore.checkPasteability()
    }

    /**
     * Key down event handler, for document-specific commands that cannot be handled via a menu accelerator.
     */
    function onKeyDown ($event: KeyboardEvent) {
      if ($event.target instanceof HTMLInputElement) return
      if (keydownAnimationFrameId) return

      keydownAnimationFrameId = requestAnimationFrame(() => {
        keydownAnimationFrameId = 0

        switch ($event.key) {
          case 'Escape':
            return store.deselectAll()
          case 'a':
            // the CommandOrControl+A accelerator doesn't work with Electron's menu
            // this is a workaround for that
            return $event.ctrlKey && store.selectAll()
          default:
            return moveItemsByArrowKey($event)
        }
      })
    }

    /**
     * Resets the arrow key momentum when the key is released.
     */
    function onKeyUp () {
      if (acceleration > 1) {
        store.commitCachedState()
        acceleration = 1
      }
    }

    /**
     * Moves the selected items by the given delta, based on the arrow key that was pressed.
     */
    function moveItemsByArrowKey ($event: KeyboardEvent) {
      const delta = (() => {
        switch ($event.key) {
          case 'ArrowLeft':
            return { x: -1, y: 0 }
          case 'ArrowUp':
            return { x: 0, y: -1 }
          case 'ArrowRight':
            return { x: 1, y: 0 }
          case 'ArrowDown':
            return { x: 0, y: 1 }
        }
      })()

      if (delta) {
        const i = Math.min(acceleration, 10)

        if (acceleration === 1) {
          store.cacheState()
        }

        store.setSelectionPosition({
          x: delta.x * i,
          y: delta.y * i
        })
        acceleration *= ARROW_KEY_MOMENTUM_MULTIPLIER
      }
    }

    /**
     * Shows the editor context menu.
     */
    function onContextMenu ($event: Event) {
      setTimeout(() => {
        window.api.showContextMenu(createEditSubmenu(props.store))
      })

      $event.preventDefault()
      $event.stopPropagation()
    }

    /**
     * Deselects all items when clicking on the background, unless the shift key is held down.
     * This is to allow a selection event to add to an existing selection by using the shift key.
     */
    function onDeselect (ctrlKey: boolean) {
      if (!ctrlKey) {
        store.deselectAll()
      }
    }

    /**
     * Prints the document using a physical printer.
     */
    async function printImage () {
      initiatePrint(async (editorElement, boundingBox, styles) => {
        await printing.printImage(editorElement, boundingBox, {
          ...styles,
          ...PRINTER_FRIENDLY_COLORS
        })

        store.status = DocumentStatus.Ready
      })
    }

    /**
     * Exports the document as an image.
     */
    async function exportImage () {
      initiatePrint(async (editorElement, boundingBox, styles) => {
        const { printArea } = printing.createPrintArea(editorElement, boundingBox, IMAGE_PADDING, styles)
        const image = await printing.createImage<Blob>(printArea, 'toBlob')

        await rootStore.saveImage(image)

        store.status = DocumentStatus.Ready
      })
    }

    /**
     * Invokes the given callback with the editor and its computed bounding box of all elements.
     * This is used for rendering the document as an image that can be printed or exported to a file.
     */
    function initiatePrint (callback: (
      editorElement: HTMLElement,
      boundingBox: BoundingBox,
      styles: Record<string, string>
    ) => Promise<void>) {
      const points = Object
        .values(store.connections)
        .reduce((points, { controlPoints }) => points.concat(
          controlPoints.map(({ position }) => position)
        ), [] as Point[])
      const boundingBoxes = Object
        .values(store.items)
        .map(({ boundingBox }) => boundingBox)
        .concat(boundaries.getConstellationBoundingBox(points))
      const boundingBox = boundaries.getGroupBoundingBox(boundingBoxes)

      callback(editorRef.value!.grid, boundingBox, {
        'zoom': `${(1 / store.zoom)}`,
        '--media-display': 'none',
        '--color-selection': 'transparent !important',
        ...preferencesStore.colorStyles
      })
    }

    return {
      store,
      editorRef,
      updates,
      gridSize,
      flash,
      storeDefinition: props.store,
      ITEM_BASE_Z_INDEX,
      onKeyDown,
      onKeyUp,
      onContextMenu,
      onDeselect
    }
  }
})
</script>
