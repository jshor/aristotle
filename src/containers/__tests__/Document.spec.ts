import { mount, VueWrapper } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import { ComponentPublicInstance } from 'vue'
import Document from '../Document.vue'
import { createDocumentStore, DocumentStore, DocumentStoreInstance } from '@/store/document'
import { createItem, createConnection, createPort, stubAll } from '@/store/document/actions/__tests__/__helpers__'
import ItemType from '@/types/enums/ItemType'
import PortType from '@/types/enums/PortType'
import printing from '@/utils/printing'
import { useRootStore } from '@/store/root'
import editorContextMenu from '@/menus/context/editor'
import { ARROW_KEY_MOMENTUM_MULTIPLIER, IMAGE_PADDING } from '@/constants'
import Connection from '@/types/interfaces/Connection'
import Port from '@/types/interfaces/Port'
import Item from '@/types/interfaces/Item'
import Point from '@/types/interfaces/Point'

setActivePinia(createPinia())

describe('Document container', () => {
  let wrapper: VueWrapper
  let storeDefinition: DocumentStore
  let store: DocumentStoreInstance
  let item1: Item, item2: Item
  let port1: Port, port2: Port
  let connection1: Connection

  beforeEach(() => {
    storeDefinition = createDocumentStore('test-id')
    store = storeDefinition()

    item1 = createItem('item1', ItemType.InputNode, { portIds: ['port1'] })
    item2 = createItem('item2', ItemType.Freeport, { portIds: ['port2'] })
    port1 = createPort('port1', item1.id, PortType.Input)
    port2 = createPort('port1', item2.id, PortType.Output)
    connection1 = createConnection('connection1', 'port1', 'port2')

    store.ports = { port1, port2 }
    store.items = { item1, item2 }
    store.connections = { connection1 }
    store.hasLoaded = true

    wrapper = mount(Document, {
      props: {
        store: storeDefinition
      }
    })

    jest
      .spyOn(window, 'requestAnimationFrame')
      .mockImplementation(cb => {
        cb(1)
        return 0
      })
  })

  afterEach(() => jest.resetAllMocks())

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when the store has not yet loaded', () => {
    it('should not render any connections or items', async () => {
      store.hasLoaded = false

      await wrapper.vm.$nextTick()

      expect(wrapper.findAll('.item')).toHaveLength(0)
      expect(wrapper.findAll('.connection')).toHaveLength(0)
    })
  })

  describe('when printing an image', () => {
    beforeEach(() => {
      stubAll(printing, ['printImage'])

      store.isPrinting = true
    })

    it('should invoke saveImage() on the root store', () => {
      expect(printing.printImage).toHaveBeenCalledWith(store.zoom, expect.any(HTMLElement), expect.anything())
    })

    it('should reset `isCreatingImage` back to `false`', () => {
      expect(store.isPrinting).toBe(false)
    })
  })

  describe('when exporting an image', () => {
    const imageData = 'image-blob-data'
    const rootStore = useRootStore()
    const printArea = document.createElement('div')

    beforeEach(() => {
      stubAll(rootStore, ['saveImage'])

      jest
        .spyOn(printing, 'createPrintArea')
        .mockReturnValue({
          printArea,
          width: 100,
          height: 100
        })
      jest
        .spyOn(printing, 'createImage')
        .mockResolvedValue(imageData)

      store.isCreatingImage = true
    })

    it('should render the print area', () => {
      expect(printing.createPrintArea).toHaveBeenCalledWith(
        1 / store.zoom,
        expect.any(HTMLElement),
        expect.anything(),
        IMAGE_PADDING,
        ''
      )
    })

    it('should render the image', () => {
      expect(printing.createImage).toHaveBeenCalledWith(printArea, 'toBlob')
    })

    it('should invoke saveImage() on the root store', () => {
      expect(rootStore.saveImage).toHaveBeenCalledWith(imageData)
    })

    it('should reset `isCreatingImage` back to `false`', () => {
      expect(store.isCreatingImage).toBe(false)
    })

    it('should not attempt to create an image if the canvas is destroyed', async () => {
      store.isCreatingImage = false

      jest.clearAllMocks()

      await wrapper.unmount()
      store.isCreatingImage = true

      expect(printing.createImage).not.toHaveBeenCalled()
    })
  })

  describe('when pressing the escape key', () => {
    it('should deselect all items', () => {
      stubAll(store, ['deselectAll'])

      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape' }))

      expect(store.deselectAll).toHaveBeenCalled()
    })
  })

  describe('when pressing \'a\'', () => {
    beforeEach(() => stubAll(store, ['selectAll']))

    it('should select all items when the control or command key is held', () => {
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'a', ctrlKey: true }))

      expect(store.selectAll).toHaveBeenCalled()
    })

    it('should not select anything when the control or command key is not held', () => {
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'a', ctrlKey: false }))

      expect(store.selectAll).not.toHaveBeenCalled()
    })

    it('should ignore any key events when the target is an input element', async() => {
      const input = document.createElement('input')
      document.body.appendChild(input)
      input.focus()

      await wrapper.vm.$nextTick()

      ;(wrapper.vm as any).onKeyDown({ key: 'a', ctrlKey: true, target: input })

      expect(store.selectAll).not.toHaveBeenCalled()
    })
  })

  describe('when pressing the arrow keys', () => {
    beforeEach(() => stubAll(store, ['moveSelectionPosition']))
    const testArrowKey = (direction: string, point: Point) => {
      it(`should invoke moveSelectionPosition() with the given ${direction} when that key is pressed`, () => {
        document.dispatchEvent(new KeyboardEvent('keydown', { key: `Arrow${direction}` }))

        expect(store.moveSelectionPosition).toHaveBeenCalledWith(point)
      })
    }

    testArrowKey('Left', { x: -1, y: 0 })
    testArrowKey('Up', { x: 0, y: -1 })
    testArrowKey('Right', { x: 1, y: 0 })
    testArrowKey('Down', { x: 0, y: 1 })

    it('should not invoke moveSelectionPosition() when a non-arrow key is pressed', () => {
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 't' }))

      expect(store.moveSelectionPosition).not.toHaveBeenCalled()
    })

    it('should reset the momentum when the key is released', () => {
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowUp' }))
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowUp' }))
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowUp' }))

      const momentum = Array(3)
        .fill(0)
        .reduce((total, i) => total * ARROW_KEY_MOMENTUM_MULTIPLIER, 1)

      expect(store.moveSelectionPosition).toHaveBeenCalledWith({ x: 0, y: -momentum })

      document.dispatchEvent(new KeyboardEvent('keyup'))
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowUp' }))

      expect(store.moveSelectionPosition).toHaveBeenCalledWith({ x: 0, y: -1 })
    })
  })

  describe('when the context menus are requested', () => {
    it('should show the editor context menu when an item requests it', async () => {
      stubAll(window.api, ['showContextMenu'])

      await wrapper
        .findComponent({ name: 'Item' })
        .vm
        .$emit('contextmenu', new Event('contextmenu'))

      expect(window.api.showContextMenu).toHaveBeenCalledTimes(1)
      expect(window.api.showContextMenu).toHaveBeenCalledWith(editorContextMenu(storeDefinition))
    })
  })

  describe('when the selector has begun a new selection', () => {
    it('should deselect all items when the shift key is not held', async () => {
      stubAll(store, ['deselectAll'])

      await wrapper
        .findComponent({ name: 'Selector' })
        .vm
        .$emit('selectionStart', {
          ctrlKey: false
        })

      expect(store.deselectAll).toHaveBeenCalled()
    })

    it('should not deselect anything when the shift key is held', async () => {
      stubAll(store, ['deselectAll'])

      await wrapper
        .findComponent({ name: 'Selector' })
        .vm
        .$emit('selectionStart', {
          ctrlKey: true
        })

      expect(store.deselectAll).not.toHaveBeenCalled()
    })
  })
})
