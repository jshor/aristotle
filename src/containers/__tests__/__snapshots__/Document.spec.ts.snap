// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document container should match the snapshot 1`] = `
"<div class="editor" tabindex="0">
  <div style="background-size: 20px 20px; transform: scale(1); left: 0px; top: 0px; width: 0px; height: 0px;" class="editor__grid">
    <!-- when tabbed into, this resets the selection back to the last focusable element in the editor -->
    <div tabindex="0" data-test="first"></div><!-- editor content (items, connections, groups, ports, etc.) -->
    <div class="selector">
      <!--v-if-->
    </div>
    <div class="draggable" style="left: 0px; top: 0px; pointer-events: none; z-index: 1000;" tabindex="0" data-test="item">
      <!--v-if-->
      <!--v-if-->
      <div class="circuit-component" style="transform: rotate(0deg);">
        <div class="switch switch--hi-z" type="InputNode" subtype="None" isselected="false" flash="false" name="" properties="[object Object]">
          <div class="switch__inner" tabindex="0" aria-role="switch">
            <div class="switch__trigger">
              <div class="switch__light"></div>
              <div class="switch__characters"></div>
            </div>
          </div>
        </div>
        <div class="circuit-component__ports--left circuit-component__ports">
          <div class="port-pivot" style="transform: rotate(-0deg);" data-port-id="port1">
            <div class="port-handle" id="port1" is-monitored="false" tabindex="0">
              <div class="port-handle__display">
                <div class="port-handle__hue"></div>
              </div>
            </div>
          </div>
        </div>
        <div class="circuit-component__ports--top circuit-component__ports"></div>
        <div class="circuit-component__ports--right circuit-component__ports"></div>
        <div class="circuit-component__ports--bottom circuit-component__ports"></div>
      </div>
    </div>
    <div class="draggable" style="left: 0px; top: 0px; pointer-events: none; z-index: 1000;" tabindex="0" data-test="item">
      <!--v-if-->
      <!--v-if-->
      <div class="circuit-component" style="transform: rotate(0deg);">
        <div class="switch switch--hi-z" type="InputNode" subtype="None" isselected="false" flash="false" name="" properties="[object Object]">
          <div class="switch__inner" tabindex="0" aria-role="switch">
            <div class="switch__trigger">
              <div class="switch__light"></div>
              <div class="switch__characters"></div>
            </div>
          </div>
        </div>
        <div class="circuit-component__ports--left circuit-component__ports">
          <div class="port-pivot" style="transform: rotate(-0deg);" data-port-id="port1">
            <div class="port-handle" id="port1" is-monitored="false" tabindex="0">
              <div class="port-handle__display">
                <div class="port-handle__hue"></div>
              </div>
            </div>
          </div>
        </div>
        <div class="circuit-component__ports--top circuit-component__ports"></div>
        <div class="circuit-component__ports--right circuit-component__ports"></div>
        <div class="circuit-component__ports--bottom circuit-component__ports"></div>
      </div>
    </div><!-- Note: z-index of items will be offset by ITEM_BASE_Z_INDEX to ensure they overlap wires -->
    <!-- the tab-accessible centroid element -->
    <div style="position: absolute; left: 0px; top: 0px;" tabindex="0" aria-label="Connection from  to " title="Connection from  to " data-test="centroid"></div><!-- each segment of the visible wire -->
    <!-- TODO: set is-animated according to preferences --><svg class="wire" width="15" height="15" aria-roledescription="connection" style="left: -7.5px; top: -7.5px;" title="Connection from  to ">
      <path class="wire__outline" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
      <path class="wire__display" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
      <!--v-if-->
      <path class="wire__clickable" fill="none" transform="translate(7.5, 7.5)" d="M 0,0 L 0,0"></path>
    </svg><!-- the visual outline of the bounding box surrounding the connection -->
    <!--v-if-->
    <!-- the control points -->
    <!-- render only n-1 control points, since the last one is the target port -->
    <!--v-if-->
    <div class="draggable" style="left: 0px; top: 0px; z-index: 1000;" tabindex="-1">
      <!--v-if-->
      <div class="port-pivot" style="transform: rotate(-0deg);">
        <div class="port-handle" data-test="port">
          <div class="port-handle__display">
            <div class="port-handle__hue"></div>
          </div>
        </div>
      </div>
    </div>
    <!--v-if-->
    <div class="draggable" style="left: 0px; top: 0px; z-index: 1000;" tabindex="-1">
      <!--v-if-->
      <div class="port-pivot" style="transform: rotate(-0deg);">
        <div class="port-handle" data-test="port">
          <div class="port-handle__display">
            <div class="port-handle__hue"></div>
          </div>
        </div>
      </div>
    </div><!-- when tabbed into, this resets the selection back to the first focusable element in the editor -->
    <div tabindex="0" data-test="last"></div>
  </div>
</div>"
`;
